feat: Allow testing next ES version during migration

luismulinari opened the pull request:
## Description

This PR introduces the ability to test Elasticsearch 8 during the ES7→ES8 migration process, enabling safer rollouts and validation before fully switching traffic. The implementation provides multiple trigger mechanisms and intelligent mirroring to ensure data consistency during testing.


### Changes Made

#### Core Functionality
- **Added ES version testing capability** via multiple trigger methods:
  - `VIP_ELASTICSEARCH_TEST_ES_NEXT` constant
  - `vip-test-es-next` cookie
  - `vip-test-es-next` query parameter
- **Enhanced URL building logic** with port-based version routing (ES7: 9234/9235, ES8: 9244/9245)
- **Intelligent mirroring** that reverses direction when testing next version
- **HTTP header injection** (`X-Elasticsearch-Version`) for debugging and monitoring

#### Code Refactoring
- **Removed migration setup from enterprise-search integration** - consolidated version logic in Search class
- **Replaced hardcoded migration endpoints** with dynamic URL building
- **Added version constants** for maintainable version references
- **Default ES version fallback** to version 7 when not explicitly set


## Pre-review checklist

Please make sure the items below have been covered before requesting a review:

- [ ] This change works and has been tested locally or in Codespaces (or has an appropriate fallback).
- [ ] This change works and has been tested on a sandbox.
- [ ] This change has relevant unit tests (if applicable).
- [ ] This change uses a rollout method to ease with deployment (if applicable - especially for large scale actions that require writes).
- [ ] This change has relevant documentation additions / updates (if applicable).
- [ ] I've created a changelog description that aligns with the provided examples.

## Pre-deploy checklist

- [ ] VIP staff: Ensure any alerts added/updated conform to internal standards (see internal documentation). 

## Steps to Test

<!--
Outline the steps to test and verify the PR here.

Example:

1. Check out PR.
1. Go to `wp-admin` > `Tools` > `Bakery`
1. Click on "Bake Cookies" button.
1. Verify cookies are delicious.
-->

luismulinari committed:
feat: Allow testing next ES version during migration

- Remove migration setup from enterprise-search integration
- Add testing capability via VIP_ELASTICSEARCH_TEST_ES_NEXT constant
- Support testing via vip-test-es-next cookie/query param
- Update mirroring logic to handle testing scenarios

luismulinari committed:
Merge branch 'develop' into es/8-migration-allow-testing-next-version

luismulinari committed:
Lint fixes

luismulinari committed:
Merge branch 'es/8-migration-allow-testing-next-version' of github.com:Automattic/vip-go-mu-plugins into es/8-migration-allow-testing-next-version

luismulinari committed:
Fix tests

luismulinari committed:
Remove temporary code

luismulinari committed:
Do not cache when testing next version

luismulinari committed:
feat: Add session-based ES version testing support

rinatkhaziev commented on search/includes/classes/class-search.php:
```diff
@@ -948,16 +957,26 @@ public function filter__ep_do_intercept_request( $request, $query, $args, $failu
+		if ( $is_testing_next_version ) {
+			$url = $this->build_elasticsearch_url( $query['url'], self::ELASTICSEARCH_MIGRATION_NEXT );
+
+			if ( ! headers_sent() ) {
+				header( sprintf( 'X-Elasticsearch-Version: %s', self::ELASTICSEARCH_MIGRATION_NEXT ) );
+			}
```
```suggestion
			add_action( 'send_headers', fn() => header( sprintf( 'X-Elasticsearch-Version: %s', self::ELASTICSEARCH_MIGRATION_NEXT ) ) );
```
WP-idiomatic way would be to use `send_headers` hook. NBD but something to be aware of.

rinatkhaziev commented on search/includes/classes/class-search.php:
```diff
@@ -1047,15 +1066,11 @@ public function filter__ep_do_intercept_request( $request, $query, $args, $failu
 	 * @param string|null $type The type of request.
 	 */
-	protected function mirror_write_to_migration_hosts( $query, $args, $type = null ) {
+	protected function mirror_write_to_migration_hosts( $query, $args, $type = null, $is_testing_next_version = false ) {
```
now missing $is_testing_next_version from docblock above

rinatkhaziev commented on search/includes/classes/class-search.php:
```diff
@@ -2500,35 +2508,124 @@ public function load_collector(): void {
+				// If the version is not 7, replace the 7 ports (when present) with the next version ports
+				$port = str_replace( $elasticsearch_seven_ports, $elasticsearch_next_ports, $port );
+			}
+		}
+
+		$new_url = '';
+
+		if ( isset( $parsed_url['scheme'] ) ) {
+			$new_url .= $parsed_url['scheme'] . '://';
+		}
+
+		$new_url .= $host;
+
+		if ( isset( $port ) ) {
+			$new_url .= sprintf( ':%d', $port );
 		}
 
-		$migration_url = $migration_host;
-		
 		if ( isset( $parsed_url['path'] ) ) {
-			$migration_url .= $parsed_url['path'];
+			$new_url .= $parsed_url['path'];
 		}
-		
+
 		if ( isset( $parsed_url['query'] ) ) {
-			$migration_url .= '?' . $parsed_url['query'];
+			$new_url .= '?' . $parsed_url['query'];
 		}
-		
+
 		if ( isset( $parsed_url['fragment'] ) ) {
-			$migration_url .= '#' . $parsed_url['fragment'];
+			$new_url .= '#' . $parsed_url['fragment'];
 		}
-		
-		return $migration_url;
+
+		return $new_url;
+	}
```
Do we really need to do this url deconstruction/reconstruction?
 Could simply be:

```php
str_replace( [ ...$old_ports ], [...$new_ports ], $url );
```

luismulinari replied:
The idea was to ensure we're correctly changing ports.

Using a `str_replace` could accidentally alter other parts of the URL unrelated to the ports, such as an index for a site with ID 9234. To prevent this, we might add `:` to the port number, but we still can't guarantee that the URL won't contain `:9234` elsewhere that isn't related to the port.

I'll try to find a better way to handle this.

luismulinari replied:
@rinatkhaziev What do you think about: https://github.com/Automattic/vip-go-mu-plugins/pull/6479/commits/8e88f2969f904a99c46d4c91a437453a44eb7cc7

rinatkhaziev reviewed (changes requested):
Left a couple of small nitpicks and a bigger one.

luismulinari committed:
Update search/includes/classes/class-search.php

Co-authored-by: Rinat K <rinat@automattic.com>

luismulinari committed:
refactor: Simplify Elasticsearch URL version replacement logic

Replace complex URL parsing and rebuilding with regex-based port replacement for better performance and maintainability. Update tests to handle URLs with paths containing port-like patterns.

luismulinari committed:
Simplify URL port change logic
